### ISSUE TITLE
[Bug]: Body -> Decorations -> Background-image

### ISSUE THREAD

**Issue by yosoyfran** [artf/grapesjs#1357]
Version 0.14.25 | Local copy GrapesJS
Tested in: Chrome 68, Firefox 61

Add a background image in "Body (Selected #wrapper) -> Decorations -> Background -> Layer # -> Image", save to db, refresh page, load from db, blank background. Same with image or color.

Adding background
![Adding background](https://user-images.githubusercontent.com/42306769/43997738-970f814a-9dba-11e8-8ec8-ffb41601cf3e.png)

After refresh page
![After refresh page](https://user-images.githubusercontent.com/42306769/43997742-ce7422a8-9dba-11e8-9d91-67aa282e03ce.png)

and, how can add a background image gif or video? need to pull a request?

Thanks.

**Comment by artf** [comment]
I need to know what do you store in DB and how do you load it.

**Comment by yosoyfran** [comment]
Hi, this is the code.

```

<div id="builder"></div>

<script type="text/javascript">

	var token = "{{ csrf_token() }}";

	var editor = grapesjs.init({
     	fromElement: 1,
     	clearOnRender: true,
     	container: '#builder',
     	plugins: ['gjs-preset-newsletter', 'gjs-blocks-basic', 'gjs-blocks-flexbox'],
     	pluginsOpts: {
			'gjs-blocks-basic': { category: false },
	    },

    	storageManager: {
        	autosave: false,
        	setStepsBeforeSave: 1,
        	type: 'remote',
        	urlStore: '/dashboard/pages/builder/' + {{ $page->id }},
        	urlLoad: '/dashboard/pages/builder/' + {{ $page->id }} + '/load',
        	contentTypeJson: true,
        	params: { _token:token },
		},

		assetManager: {
			assets: {!! !empty($page->builder['gjs-assets']) ? $page->builder['gjs-assets'] : '[]' !!},
			storeOnChange : true,
			storeAfterUpload : true,
		    upload: '/dashboard/pages/builder/' + {{ $page->id }} + '/postimages',
		    uploadName: 'images',
		    params: { _token:token, enctype:'multipart/form-data' },
		    autoAdd: 1,
		},
	});

	var assets = editor.AssetManager.getAll();

	assets.on('remove', function(asset) {

		var removefile = asset.get('src').split('/').reverse()[0];

        $.ajax({
        	url: '/dashboard/pages/builder/' + {{ $page->id }} + '/postimages',
        	type: 'POST',
        	data: { file:removefile, _token:token },
        	success: function(response){
        		assets.remove(asset.get('src'));
        	}
    	});
	});

	editor.Panels.addButton
	('options',
		[{
    		id: 'save-db',
    		className: 'fa fa-floppy-o',
    		command: 'save-db',
    		attributes: {title: 'Save'}
        }]
	);

	editor.Commands.add
    ('save-db',
	    {
	        run: function(editor, sender)
	        {
	        	sender && sender.set('active');
	        	editor.store();
        	}
    });

        editor.Panels.removeButton("options", ["export-template", "gjs-toggle-images", "gjs-open-import-template"]);

    editor.on('storage:load', function(e) {
        console.log('Loaded ', e);
	});

	editor.on('storage:store', function(e) {
        console.log('Stored ', e);
	});

	var btnClass = editor.Panels.addButton
	('options',
		[{
    		id: 'add-class',
    		className: 'fa fa-plus',
    		command: 'add-class',
    		attributes: {title: 'Add Clicker Zone'}
        }]
	);

	var cmdClass = editor.Commands.add
    ('add-class',
	    {
	        run: function(editor, sender)
	        {
	        	var componentModel = editor.getSelected();
				componentModel.addClass('open-box');
	        	sender && sender.set('active', true);
        	},

        	stop: function(editor, sender)
        	{
        		var componentModel = editor.getSelected();
        		componentModel.removeClass('open-box');
	        	sender && sender.set('active', false);
        	}
    	}
    );


</script>


```
This is the php code: store
```
public function builderPost(Page $page, Request $request)
    {
        if (!empty($page->exists) AND Gate::allows('page-access', $page)) {

            $page->builder = json_encode($request->except('_token'));
            $page->save();

            return ['status' => 'success', 'message' => 'Your Page was successfully saved!'];
        }

        Session::flash('flashSession', ['status' => 'danger', 'message' => 'Error, you cannot build this Page!']);
        return redirect('dashboard/pages');
    }
```
for load
```
public function builderLoad(Page $page)
    {
        if (!empty($page->exists) AND Gate::allows('page-access', $page)) {

            return json_decode($page->builder, true);
        }

        Session::flash('flashSession', ['status' => 'danger', 'message' => 'Error, you cannot build this Page!']);
        return redirect('dashboard/pages');
        
    }
```

Thanks.

**Comment by artf** [comment]
@yosoyfran ok but there is one thing that is not clear for me, is it just about the background or all of your content? I mean, if you add some block inside, save and refresh, you see them correctly, right? If you completely don't see anything of your content, check the load request via inspector

**Comment by yosoyfran** [comment]
@artf This problem is only in the "body" (Selected #wrapper) background, the content in the builder is ok, please check this [video gift link ](http://recordit.co/ooiazPsMoR)

Would be awesome if you add the options, Decorations -> background -> add layer -> [image, video, color], opacity/transparency. (for ex: i wanna add a video (layer 1) and color (layer 2) with transparency in front, for effect).

Thank you for your time.

**Comment by v8jupiter** [comment]
Have same issue.

**Comment by artf** [comment]
Ok thanks for the video @yosoyfran I'll try to figure it out was is going on there

**Comment by ageir** [comment]
I'm having the exact same problem. The background isn't reloaded. But the data sent back from php to the load function is correct and contains the background styles. All other content are preserved and ok.

Also when clearing the canvas (with the trashcan icon) it doesn't clear the background styles, i think that might be related. You can test it on https://grapesjs.com/demo.html Set a background image and press clear canvas. The background is still there.

**Comment by artf** [comment]
> Also when clearing the canvas (with the trashcan icon) it doesn't clear the background styles, i think that might be related. You can test it on grapesjs.com/demo.html Set a background image and press clear canvas. The background is still there.

mmmm I'd say it's just because the clear canvas command doesn't touch wrapper's style (it's about adding the line suggested by @maryia-kabash in [CanvasClear](https://github.com/artf/grapesjs/blob/dev/src/commands/view/CanvasClear.js), could be an easy PR), so not related

BTW I've added this issue for the [Next release](https://github.com/artf/grapesjs/projects/2)

**Comment by artf** [comment]
Guys, can you try to use this options and let me know if it fixes the issue:
```js
grapesjs.init({
	//...
	domComponents: { storeWrapper: 1 },
})
```

**Comment by artf** [comment]
@yosoyfran can you check what do you load? The main difference with that options is that the stored `components` key is not an array of components anymore (`[{type: 'header', ...}, {...}]`) but a component itself, the wrapper (`{ type: 'wrapper', components: [{type: 'header', ...}, {...}], ... }`)

**Comment by v8jupiter** [comment]
When template loaded again to editor no content inside only background displayed.
storeWrapper - enabled 
 ` grapesjs.init({
//...
    domComponents: { storeWrapper: 1 },`

Please look at this video - https://youtu.be/4lF4wdxfI0I

data from db: 
`{"gjs-assets":"[]","gjs-css":"* { box-sizing: border-box; } body {margin: 0;}body{background-color:#d5bebe;}.c862{padding:10px;}","gjs-styles":"[{\"selectors\":[{\"name\":\"c843\",\"label\":\"c843\",\"type\":1,\"active\":true,\"private\":false,\"protected\":false}],\"selectorsAdd\":\"\",\"style\":{\"padding\":\"10px\"},\"mediaText\":\"\",\"state\":\"\",\"stylable\":true,\"atRuleType\":\"\",\"singleAtRule\":0,\"important\":0},{\"selectors\":[{\"name\":\"c862\",\"label\":\"c862\",\"type\":1,\"active\":true,\"private\":false,\"protected\":false}],\"selectorsAdd\":\"\",\"style\":{\"padding\":\"10px\"},\"mediaText\":\"\",\"state\":\"\",\"stylable\":true,\"atRuleType\":\"\",\"singleAtRule\":0,\"important\":0}]","gjs-html":"<div class=\"c862\"><p>Insert your text here</p></div>","gjs-components":"{\"tagName\":\"div\",\"type\":\"\",\"name\":\"\",\"removable\":false,\"draggable\":false,\"droppable\":true,\"badgable\":true,\"stylable\":[\"background\",\"background-color\",\"background-image\",\"background-repeat\",\"background-attachment\",\"background-position\",\"background-size\"],\"stylable-require\":\"\",\"unstylable\":\"\",\"highlightable\":true,\"copyable\":false,\"resizable\":false,\"editable\":false,\"layerable\":true,\"selectable\":true,\"hoverable\":true,\"void\":false,\"state\":\"\",\"status\":\"selected\",\"content\":\"\",\"icon\":\"\",\"style\":{\"background-color\":\"#d5bebe\"},\"classes\":[],\"script\":\"\",\"attributes\":{\"id\":\"wrapper\"},\"traits\":[],\"propagate\":\"\",\"components\":[{\"tagName\":\"div\",\"type\":\"text\",\"name\":\"\",\"removable\":true,\"draggable\":true,\"droppable\":false,\"badgable\":true,\"stylable\":true,\"stylable-require\":\"\",\"unstylable\":\"\",\"highlightable\":true,\"copyable\":true,\"resizable\":false,\"editable\":true,\"layerable\":true,\"selectable\":true,\"hoverable\":true,\"void\":false,\"state\":\"\",\"status\":\"\",\"content\":\"<p>Insert your text here</p>\",\"icon\":\"\",\"style\":{},\"classes\":[{\"name\":\"c862\",\"label\":\"c862\",\"type\":1,\"active\":true,\"private\":false,\"protected\":false}],\"script\":\"\",\"attributes\":{},\"traits\":[{\"type\":\"text\",\"label\":\"\",\"name\":\"id\",\"min\":\"\",\"max\":\"\",\"unit\":\"\",\"step\":1,\"value\":\"\",\"default\":\"\",\"placeholder\":\"eg. Text here\",\"changeProp\":0,\"options\":[]},{\"type\":\"text\",\"label\":\"\",\"name\":\"title\",\"min\":\"\",\"max\":\"\",\"unit\":\"\",\"step\":1,\"value\":\"\",\"default\":\"\",\"placeholder\":\"eg. Text here\",\"changeProp\":0,\"options\":[]}],\"propagate\":\"\",\"activeOnRender\":0,\"components\":[]}],\"custom-name\":\"Body\",\"wrapper\":1,\"open\":0}"}`

**Comment by artf** [comment]
Thanks @v8jupiter for the video and the data, it helps a lot, but still not getting why it's not working on your machine (the data stored seems to be correct though). Do you use the latest version? How do you load the stored data, simply via `config.storageManager.urlLoad`?
For anyone who is interested, this is where we get the loaded data and check if it is an object: https://github.com/artf/grapesjs/blob/dev/src/dom_components/index.js#L327-L343

**Comment by artf** [comment]
Example with localStorage (the process is the same as the remote one) https://jsfiddle.net/n2a3rvct/ the load works correctly. Can someone confirm that it works with the **latest** version? @yosoyfran @v8jupiter @maryia-kabash @ageir

**Comment by v8jupiter** [comment]
I will check tomorrow . Need to check with latest build from dev or master ?

**Comment by artf** [comment]
@v8jupiter weird... can you post your config regarding the storage?

**Comment by v8jupiter** [comment]
`    storageManager: {
      autoload:  true,
      autosave: false,
      setStepsBeforeSave: 1,
      type: 'remote',
      urlStore: urlStore, // var from PHP
      urlLoad: loadUrl, // var from PHP
      contentTypeJson: true
    },
    assetManager: {
      upload: urlAssets, // var from PHP
    },`

**Comment by artf** [comment]
Hi @v8jupiter (and others), can you check please the new released RC https://github.com/artf/grapesjs/tree/15-beta/dist (npm https://www.npmjs.com/package/grapesjs/v/0.15.3-rc)
Unfortunately, I'm stuck with releasing the [new version](https://github.com/artf/grapesjs/projects/2) due to this bug, so if you're able to test and report (`domComponents.storeWrapper` is `true` by default now) I'll be really gratefull

**Comment by yosoyfran** [comment]
Hi @artf , i don't know if i am implement wrong or this is not working for me, this is my code
https://jsfiddle.net/jkbr16aL/

Sorry for the delay.

**Comment by artf** [comment]
@yosoyfran I didn't get if you use the 0.15.3-rc version

**Comment by v8jupiter** [comment]
@artf Tested and for me  0.15.3-rc  work as expected. Thank you! I checked with old templates and it works too.

**Comment by v8jupiter** [comment]
I found in 0.15.3-rc another issue. Not related with background, but please look https://youtu.be/uE3aNxDeaxQ

I can to reproduce only in new version

**Comment by yosoyfran** [comment]
Hi @artf, sorry i tried now in 0.15, and is working fine, thank you for your amazing work!!!

**Comment by artf** [comment]
@v8jupiter can you confirm that this happens after the load (I'm not able to reproduce it with an empty template)? In that case, can you post the `components` JSON you get from the load?

**Comment by v8jupiter** [comment]
`{"gjs-assets":"[]","gjs-css":"* { box-sizing: border-box; } body {margin: 0;}body{background-color:rgba(201,0,0,0.77);}","gjs-styles":"[]","gjs-html":"","gjs-components":"{\"tagName\":\"div\",\"type\":\"wrapper\",\"name\":\"\",\"removable\":false,\"draggable\":false,\"droppable\":true,\"badgable\":true,\"stylable\":[\"background\",\"background-color\",\"background-image\",\"background-repeat\",\"background-attachment\",\"background-position\",\"background-size\"],\"stylable-require\":\"\",\"style-signature\":\"\",\"unstylable\":\"\",\"highlightable\":true,\"copyable\":false,\"resizable\":false,\"editable\":false,\"layerable\":true,\"selectable\":true,\"hoverable\":true,\"void\":false,\"state\":\"\",\"status\":\"selected\",\"content\":\"\",\"icon\":\"\",\"style\":{\"background-color\":\"rgba(201,0,0,0.77)\"},\"classes\":[],\"script\":\"\",\"attributes\":{\"id\":\"wrapper\"},\"traits\":[],\"propagate\":\"\",\"components\":[],\"custom-name\":\"Body\",\"wrapper\":1,\"open\":false}"}`

It seems only after load and with openned debug toolbar
Steps for reproduce:
1.  Before load open dubug toolbar and make size of toolbar as on prev. video
2.  Try to move any element . And it will added before you "release" the item

**Comment by artf** [comment]
> a lot of bugs

Be more specific, please...

**Comment by v8jupiter** [comment]
Tomorrow i will retest and give you full feedback.

**Comment by artf** [comment]
Hi guys, I just published the new release https://github.com/artf/grapesjs/releases/tag/v0.14.40
which was in pending for too much. To avoid breaking changes I rolled back the default of `storeWrapper` to `false`
@v8jupiter I close this issue. For your new feedbacks please open a new one
